Remove Node dependencies from operator.h

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32761 Remove ImplicitTensorToNum
* #32760 remove unnecessary Node* ops
* **#32682 Remove Node dependencies from operator.h**

This moves code around so that operator.h/cpp no longer requires a full
definition of Node* nor does it include alias analysis or the pretty printer.

This should make it possible to include in the mobile build.

Functionality for checking if operators match Node and to look up
and operator for a Node have moved to the Node object.

Differential Revision: [D19615386](https://our.internmc.facebook.com/intern/diff/D19615386)

zdevito committed:
Remove Node dependencies from operator.h

This moves code around so that operator.h/cpp no longer requires a full
definition of Node* nor does it include alias analysis or the pretty printer.

This should make it possible to include in the mobile build.

Functionality for checking if operators match Node and to look up
and operator for a Node have moved to the Node object.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 86d3dfcc:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=f6c46df856d0588360db5b807960d1fc5e888c36&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base f6c46df8 since Jan 27
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=86d3dfcca7eb9f886d77ec08d3df062f4bb80534).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=91738810) from Jan 27 until Jan 27



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito committed:
Update on "Remove Node dependencies from operator.h"


This moves code around so that operator.h/cpp no longer requires a full
definition of Node* nor does it include alias analysis or the pretty printer.

This should make it possible to include in the mobile build.

Functionality for checking if operators match Node and to look up
and operator for a Node have moved to the Node object.

Differential Revision: [D19615386](https://our.internmc.facebook.com/intern/diff/D19615386)

[ghstack-poisoned]

zdevito committed:
Update on "Remove Node dependencies from operator.h"


This moves code around so that operator.h/cpp no longer requires a full
definition of Node* nor does it include alias analysis or the pretty printer.

This should make it possible to include in the mobile build.

Functionality for checking if operators match Node and to look up
and operator for a Node have moved to the Node object.

Differential Revision: [D19615386](https://our.internmc.facebook.com/intern/diff/D19615386)

[ghstack-poisoned]

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -853,10 +854,54 @@ Value* Node::namedInput(Symbol name) const {
+  }
+
+  // too many inputs
+  if (!schema.is_vararg() && actuals.size() != formals.size()) {
```
if schema is vararg, do extra actuals go unchecked?

zdevito replied:
yes, we check the formal arguments, then we allow any additional actual arguments.

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -853,10 +854,54 @@ Value* Node::namedInput(Symbol name) const {
+  return true;
+}
+
 bool Node::matches(
```
I know this function isn't part of this work, but is it worth adding a comment header here? Schema matching is obvious enough but it's not immediately clear why we're doing this other locally-unrelated thing on `const_inputs` at the same time (also! `should_be_const_inputs` per this afternoon :P)

No need for diff creep but thinking out loud as I read through this

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -897,20 +942,51 @@ const FunctionSchema* Node::maybeSchema() const {
+    return *maybe;
+
+  auto er = script::ErrorReport(sourceRange());
+  er << "Schema not found for node. File a bug report.\n";
```
Please?

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -897,20 +942,51 @@ const FunctionSchema* Node::maybeSchema() const {
+  if (maybe)
+    return *maybe;
+
+  auto er = script::ErrorReport(sourceRange());
```
nit: matter of taste I guess, but I'd put the error builder in its own function rather than having it be like 80% of `getOperator()`. OK zooming out now

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -897,20 +942,51 @@ const FunctionSchema* Node::maybeSchema() const {
+}
+
+Operation Node::getOperation() const {
+  // note: getOperatorFor ensures that getOperatorFor(node).matches(node) ==
```
This comment is either stale or way too subtle for me

bhosmer commented on torch/csrc/jit/ir.cpp:
```diff
@@ -941,7 +1017,7 @@ bool Node::isNondeterministic() const {
       "aten::randperm(int n, *, int? dtype, int? layout, Device? device, bool? pin_memory) -> Tensor"};
 
-  if (nondeterministic_ops.find(this) == nullptr) {
+  if (!isMemberOf(nondeterministic_ops)) {
```
OT: how would we find out if a sig change in `native_functions.yaml` caused one of these to start missing? Also, will a new op with nondeterministic behavior need to know this exists?

Simple answer would be to make this a property of the native_ops decl (or whatever single source of truth we converge on)... ?

zdevito replied:
These signatures are checked against registered ops. If we don't fine them, then there will be an error on program initialization. So typically when one edited native_ops, they quickly find that these need to get updated as well. The harder thing to detect is if someone adds a non-deterministic op.

bhosmer commented on torch/csrc/jit/operator.cpp:
```diff
@@ -131,8 +128,132 @@ OperatorRegistry& getRegistry() {
+  using namespace at;
+  // WARNING: by adding a value to this set, you are asserting
+  // that you have also added special handling of this symbol to
+  // the printer above. Not adding handling will cause import and export
```
"the printer above"

bhosmer reviewed (approved):
LGTM, just a couple nits and questions.

zdevito committed:
Update on "Remove Node dependencies from operator.h"


This moves code around so that operator.h/cpp no longer requires a full
definition of Node* nor does it include alias analysis or the pretty printer.

This should make it possible to include in the mobile build.

Functionality for checking if operators match Node and to look up
and operator for a Node have moved to the Node object.

Differential Revision: [D19615386](https://our.internmc.facebook.com/intern/diff/D19615386)

[ghstack-poisoned]
